### PR TITLE
SlackNLPModel consistency and hess_coord consistency

### PR DESCRIPTION
**Pull request by abelsiqueira**

### CODE REVIEW

**Comment by abelsiqueira** [comment]
Closes #32

File: src/simple_model.jl [hunk 1]
```diff
@@ -96,7 +96,14 @@ end
...
 function grad!(nlp :: SimpleNLPModel, x :: Vector, g :: Vector)
   nlp.counters.neval_grad += 1
-  return ForwardDiff.gradient!(g, nlp.f, x)
+  if length(x) == length(g)
+    return ForwardDiff.gradient!(g, nlp.f, x)
+  else
+    gx = zeros(length(x))
+    ForwardDiff.gradient!(gx, nlp.f, x)
+    g[1:length(x)] = gx
+    return g
```
Comment (dpo) [inline src/simple_model.jl]: I wonder if this can be resolved with array views?!

Comment (abelsiqueira) [inline src/simple_model.jl]: Thanks, I updated it with SubArray (ArrayView's sugestion for 0.4)

File: src/simple_model.jl [hunk 2]
```diff
@@ -99,9 +99,8 @@ function grad!(nlp :: SimpleNLPModel, x :: Vector, g :: Vector)
   if length(x) == length(g)
     return ForwardDiff.gradient!(g, nlp.f, x)
   else
-    gx = zeros(length(x))
+    gx = sub(g, 1:length(x))
     ForwardDiff.gradient!(gx, nlp.f, x)
-    g[1:length(x)] = gx
```
Comment (dpo) [inline src/simple_model.jl]: Is the `if ... else` still needed, or can we always do `ForwardDiff.gradient!(sub(g, 1:length(x)), nlp.f, x)`?

Comment (abelsiqueira) [inline src/simple_model.jl]: You're right, the sub call is inexpensive.

File: src/simple_model.jl [hunk 3]
```diff
@@ -128,7 +128,7 @@ end
 
 function jprod!(nlp :: SimpleNLPModel, x :: Vector, v :: Vector, Jv :: Vector)
   nlp.counters.neval_jprod += 1
-  Jv[:] = ForwardDiff.jacobian(nlp.c, x) * v
+  Jv[1:nlp.meta.ncon] = ForwardDiff.jacobian(nlp.c, x) * v
```
Comment (dpo) [inline src/simple_model.jl]: So now I'm wondering if this shouldn't remain `Jv[:]` and if the caller should use `sub(...)`. What do you think?

Comment (abelsiqueira) [inline src/simple_model.jl]: Speedwise is the same, and this (should) break less, so maybe this way is better.
If we change this back, we have to change Ampl back too.

Comment (dpo) [inline src/simple_model.jl]: Yes, and JuMP. I guess this is safe, so let's leave it.

File: src/simple_model.jl [hunk 4]
```diff
@@ -155,6 +155,17 @@ function hess(nlp :: SimpleNLPModel, x :: Vector; obj_weight = 1.0, y :: Vector
...
 
+function hess_coord(nlp :: SimpleNLPModel, x :: Vector; obj_weight = 1.0, y :: Vector = [])
+  Hx = sparse(hess(nlp, x, obj_weight=obj_weight, y=y))
+  rows = rowvals(Hx)
+  vals = nonzeros(Hx)
+  cols = Int[]
+  for i = 1:nlp.meta.nvar
+    append!(cols, i*ones(Int, length(nzrange(Hx,i))))
+  end
+  return rows, cols, vals
```
Comment (dpo) [inline src/simple_model.jl]: Doesn't `findnz()` work here?

Comment (abelsiqueira) [inline src/simple_model.jl]: Yes, thank you

Comment (dpo) [inline src/simple_model.jl]: It's too bad ForwardDiff only does dense Hessians. ReverseDiffSparse seems quite involved.

Comment (abelsiqueira) [inline src/simple_model.jl]: Yes, I hope to take a look at it, eventually. And the Autograd.jl too.

**Comment by dpo** [comment]
👍

**Pull request merged**